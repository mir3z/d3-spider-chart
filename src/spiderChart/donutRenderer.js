import * as d3 from "d3";
import svgPath from "smart-svg-path";
import takeRightWhile from "lodash.takerightwhile";

/**
 * Renders donut around the chart.
 * Each donut slice denotes one group and contains group name.
 * Each axis is marked by small circle at the edge of the donut.
 * @param parent SVG parent group
 * @param innerRadius inner radius of the donut
 * @param outerRadius outer radius of the donut
 * @returns {function}
 */
export default function donutRenderer(parent, { innerRadius, outerRadius }) {
    const groupColor = d3.scaleLinear()
        .domain([0, 4, 5])
        .range([
            "#b62b3d",
            "#ffc392",
            "#c1c8c1"
        ])
        .interpolate(d3.interpolateHcl);

    const arc = d3.arc()
        .innerRadius(innerRadius)
        .outerRadius(outerRadius);

    const group = parent
        .append("g")
        .attr("class", "donut");

    const renderDots = (axisList, groupList) => {
        const axisToGroup = axisList
            .reduce((acc, it) => ({
                ...acc,
                [it.axis]: {
                    group: it.group,
                    idx: groupList.findIndex(g => g.group === it.group)
                }
            }), {});

        const x = (i, r) => r * Math.sin(i * Math.PI * 2 / axisList.length);
        const y = (i, r) => -r * Math.cos(i * Math.PI * 2 / axisList.length);

        const dot = group
            .selectAll(`.${ CLASS_NAME.DONUT_DOT }`)
            .data(axisList);

        dot.enter()
            .append("circle")
            .attr("class", CLASS_NAME.DONUT_DOT)
            .call(setDot(x, y, axisToGroup))
            .attr("r", 5);

        dot.call(setDot(x, y, axisToGroup));

        dot.exit().remove();
    };

    const setDot = (x, y, axisToGroup) => el => el
        .attr("cx", (d, i) => x(i, outerRadius))
        .attr("cy", (d, i) => y(i, outerRadius))
        .attr("r", 5)
        .attr("fill", d => d.highlight ? "#000" : groupColor(axisToGroup[d.axis].idx));

    const setSlicePath = (arc, color) => el => el
        .attr("d", arc)
        .attr("fill", (d, i) => color(i));

    const setSliceTextPath = (arc) => el => el
        .attr("id", (d, i) => "donut-text-path-"+i)
        .attr("d", d => {
            const path = arc(d);
            // Extract the first (outer) segment of donut's slice path and use it as text path for group name
            const segments = svgPath.getCommands(path);
            const firstSegment = segments.slice(0, 2).join("");
            // Reverse path if text is upside down
            return shouldFlipText(d)
                ? svgPath.reverse(firstSegment)
                : firstSegment;
        });

    const setSliceLabel = () => el => {
        const halfDonutSize = (outerRadius - innerRadius) / 2;
        const textOffset = 4;

        el.attr("dy", d => shouldFlipText(d) ? -halfDonutSize + textOffset : halfDonutSize + textOffset);

        let textPath = el.select("textPath");

        if (textPath.size() === 0) {
            textPath = el.append("textPath");
        }

        textPath
            .attr("startOffset", "50%")
            .attr("text-anchor", "middle")
            .attr("xlink:href", (d, i) => `#donut-text-path-${ i }`)
            .text(d => d.data.group);

        return textPath;
    };

    return function render(data = []) {
        const axisList =  data[0].values;

        const groupSize = axisList
            .reduce((acc, it) => ({
                ...acc,
                [it.group]: (acc[it.group] || 0) + 1
            }), {});

        const groupList = Object.keys(groupSize)
            .map(group => ({ group, size: groupSize[group] }));

        const angleOffset = calculateDonutAngleOffset(data);

        const pie = d3.pie()
            .value(d => d.size)
            .startAngle(angleOffset)
            .endAngle(angleOffset + Math.PI * 2)
            .sort(null);

        const donutData = pie(groupList);

        renderDots(axisList, groupList);

        const sliceGroup = group
            .selectAll(`.${ CLASS_NAME.DONUT_SLICE_GROUP }`)
            .data(donutData);

        const enterSlice = sliceGroup
            .enter()
            .append("g")
            .attr("class", CLASS_NAME.DONUT_SLICE_GROUP);

        sliceGroup.exit().remove();

        enterSlice
            .append("path")
            .attr("class", CLASS_NAME.DONUT_SLICE)
            .call(setSlicePath(arc, groupColor));

        sliceGroup
            .select(`.${ CLASS_NAME.DONUT_SLICE }`)
            .call(setSlicePath(arc, groupColor));

        enterSlice
            .append("path")
            .attr("class", CLASS_NAME.DONUT_TEXT_PATH)
            .style("fill", "none")
            .call(setSliceTextPath(arc));

        sliceGroup
            .select(`.${ CLASS_NAME.DONUT_TEXT_PATH }`)
            .call(setSliceTextPath(arc));

        enterSlice
            .append("text")
            .attr("class", CLASS_NAME.DONUT_LABEL)
            .call(setSliceLabel());

        sliceGroup
            .select(`.${ CLASS_NAME.DONUT_LABEL }`)
            .call(setSliceLabel());

        return group;
    };
}

const shouldFlipText = arc => {
    const { startAngle, endAngle } = arc;
    const angle = startAngle + (endAngle - startAngle) / 2;
    return angle > Math.PI / 2 && angle < Math.PI * 3/2;
};

const calculateDonutAngleOffset = data => {
    const shift = 2 * Math.PI / data[0].values.length;
    let n = 0;
    if (data[0].values[0].group === data[0].values[data[0].values.length - 1].group) {
        n = takeRightWhile(data[0].values, ({ group }) => group === data[0].values[0].group).length;
    }
    return -(n + 0.5) * shift;
};

const CLASS_NAME = {
    DONUT_DOT: "donut-dot",
    DONUT_SLICE_GROUP: "donut-slice-group",
    DONUT_SLICE: "donut-slice",
    DONUT_TEXT_PATH: "donut-text-path",
    DONUT_LABEL: "donut-label"
};
