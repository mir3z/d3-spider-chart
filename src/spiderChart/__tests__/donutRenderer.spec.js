import donutRenderer from "../donutRenderer";
import { cleanup, setup } from "./utils";


describe("donut renderer", () => {
    const defaultOptions = { innerRadius: 100, outerRadius: 130 };
    const data = [{
        values: [
            { group: "Group A" },
            { group: "Group A" },
            { group: "Group B" },
            { group: "Group B", highlight: true },
            { group: "Group B" },
            { group: "Group B" },
            { group: "Group C" },
            { group: "Group C" },
            { group: "Group D" },
            { group: "Group D" },
            { group: "Group D" },
            { group: "Group E" }
        ]
    }];

    afterEach(cleanup);

    it("renders its own container", () => {
        const svg = setup();
        const renderDonut = donutRenderer(svg, defaultOptions);

        const group = renderDonut(data);

        const ownGroup = svg.node().querySelector("g.donut");
        expect(ownGroup).not.toBeNull();
        expect(group.node()).toEqual(ownGroup);
    });

    it("correctly renders donut slice for each group", () => {
        const svg = setup();
        const renderDonut = donutRenderer(svg, defaultOptions);

        const group = renderDonut(data);

        expect(group.node()).toMatchSnapshot();
    });

    it("updates donut when data changes", () => {
        const svg = setup();
        const renderDonut = donutRenderer(svg, defaultOptions);

        renderDonut(data);
        const group = renderDonut([{
            values: [
                { group: "Group A", highlight: true },
                { group: "Group A" },
                { group: "Group B" },
                { group: "Group B" },
                { group: "Group B" },
                { group: "Group B" }
            ]
        }]);

        expect(group.node()).toMatchSnapshot();
    });
});
