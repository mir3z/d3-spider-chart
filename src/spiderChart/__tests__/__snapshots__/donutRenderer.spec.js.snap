// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`donut renderer correctly renders donut slice for each group 1`] = `
<g
  class="donut"
>
  <circle
    class="donut-dot"
    cx="0"
    cy="-130"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="64.99999999999999"
    cy="-112.58330249197704"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="112.58330249197702"
    cy="-65.00000000000001"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="130"
    cy="-7.960204194457796e-15"
    fill="#000"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="112.58330249197704"
    cy="64.99999999999997"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="64.99999999999999"
    cy="112.58330249197704"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="1.592040838891559e-14"
    cy="130"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-64.99999999999996"
    cy="112.58330249197705"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-112.58330249197701"
    cy="65.00000000000006"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-130"
    cy="2.3880612583373385e-14"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-112.58330249197702"
    cy="-65.00000000000001"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-65.00000000000006"
    cy="-112.583302491977"
    fill="rgb(255, 195, 146)"
    r="5"
  />
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M-33.64647586332768,-125.57035741757888A130,130,0,0,1,91.92388155425118,-91.92388155425117L70.71067811865476,-70.71067811865474A100,100,0,0,0,-25.881904510252063,-96.59258262890683Z"
      fill="rgb(182, 43, 61)"
    />
    <path
      class="donut-text-path"
      d="M-33.64647586332768 -125.57035741757888 A130 130 0 0 1 91.92388155425118 -91.92388155425117 "
      id="donut-text-path-0"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="19"
    >
      <textpath
        href="#donut-text-path-0"
        startOffset="50%"
        text-anchor="middle"
      >
        Group A
      </textpath>
    </text>
  </g>
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M91.92388155425118,-91.92388155425117A130,130,0,0,1,33.64647586332772,125.57035741757886L25.881904510252095,96.59258262890683A100,100,0,0,0,70.71067811865476,-70.71067811865474Z"
      fill="rgb(202, 87, 75)"
    />
    <path
      class="donut-text-path"
      d="M 33.64647586332772 125.57035741757886 A 130 130 0 0 0 91.92388155425118 -91.92388155425117"
      id="donut-text-path-1"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="-11"
    >
      <textpath
        href="#donut-text-path-1"
        startOffset="50%"
        text-anchor="middle"
      >
        Group B
      </textpath>
    </text>
  </g>
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M33.64647586332772,125.57035741757886A130,130,0,0,1,-91.92388155425117,91.92388155425118L-70.71067811865474,70.71067811865476A100,100,0,0,0,25.881904510252095,96.59258262890683Z"
      fill="rgb(221, 124, 93)"
    />
    <path
      class="donut-text-path"
      d="M -91.92388155425117 91.92388155425118 A 130 130 0 0 0 33.64647586332772 125.57035741757886"
      id="donut-text-path-2"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="-11"
    >
      <textpath
        href="#donut-text-path-2"
        startOffset="50%"
        text-anchor="middle"
      >
        Group C
      </textpath>
    </text>
  </g>
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M-91.92388155425117,91.92388155425118A130,130,0,0,1,-91.9238815542512,-91.92388155425117L-70.71067811865477,-70.71067811865474A100,100,0,0,0,-70.71067811865474,70.71067811865476Z"
      fill="rgb(238, 160, 117)"
    />
    <path
      class="donut-text-path"
      d="M-91.92388155425117 91.92388155425118 A130 130 0 0 1 -91.9238815542512 -91.92388155425117 "
      id="donut-text-path-3"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="19"
    >
      <textpath
        href="#donut-text-path-3"
        startOffset="50%"
        text-anchor="middle"
      >
        Group D
      </textpath>
    </text>
  </g>
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M-91.9238815542512,-91.92388155425117A130,130,0,0,1,-33.64647586332768,-125.57035741757888L-25.881904510252063,-96.59258262890683A100,100,0,0,0,-70.71067811865477,-70.71067811865474Z"
      fill="rgb(255, 195, 146)"
    />
    <path
      class="donut-text-path"
      d="M-91.9238815542512 -91.92388155425117 A130 130 0 0 1 -33.64647586332768 -125.57035741757888 "
      id="donut-text-path-4"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="19"
    >
      <textpath
        href="#donut-text-path-4"
        startOffset="50%"
        text-anchor="middle"
      >
        Group E
      </textpath>
    </text>
  </g>
</g>
`;

exports[`donut renderer updates donut when data changes 1`] = `
<g
  class="donut"
>
  <circle
    class="donut-dot"
    cx="0"
    cy="-130"
    fill="#000"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="112.58330249197702"
    cy="-65.00000000000001"
    fill="rgb(202, 87, 75)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="112.58330249197704"
    cy="64.99999999999997"
    fill="rgb(202, 87, 75)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="1.592040838891559e-14"
    cy="130"
    fill="rgb(202, 87, 75)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-112.58330249197701"
    cy="65.00000000000006"
    fill="rgb(202, 87, 75)"
    r="5"
  />
  <circle
    class="donut-dot"
    cx="-112.58330249197702"
    cy="-65.00000000000001"
    fill="rgb(202, 87, 75)"
    r="5"
  />
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M-64.99999999999997,-112.58330249197704A130,130,0,0,1,130,0L100,0A100,100,0,0,0,-49.99999999999998,-86.60254037844388Z"
      fill="rgb(182, 43, 61)"
    />
    <path
      class="donut-text-path"
      d="M-64.99999999999997 -112.58330249197704 A130 130 0 0 1 130 0 "
      id="donut-text-path-0"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="19"
    >
      <textpath
        href="#donut-text-path-0"
        startOffset="50%"
        text-anchor="middle"
      >
        Group A
      </textpath>
    </text>
  </g>
  <g
    class="donut-slice-group"
  >
    <path
      class="donut-slice"
      d="M130,0A130,130,0,1,1,-65.00000000000006,-112.58330249197701L-50.00000000000004,-86.60254037844385A100,100,0,1,0,100,0Z"
      fill="rgb(202, 87, 75)"
    />
    <path
      class="donut-text-path"
      d="M -65.00000000000006 -112.58330249197701 A 130 130 0 1 0 130 0"
      id="donut-text-path-1"
      style="fill: none;"
    />
    <text
      class="donut-label"
      dy="-11"
    >
      <textpath
        href="#donut-text-path-1"
        startOffset="50%"
        text-anchor="middle"
      >
        Group B
      </textpath>
    </text>
  </g>
</g>
`;
